standard/statements.md

```diff
@@ -624,93 +622,87 @@ Multiple labels are permitted in a *switch_section*.
 > *end example*
 <!-- markdownlint-disable MD028 -->
 
-<!-- markdownlint-enable MD028 -->
-> *Note*: The “no fall through” rule prevents a common class of bugs that occur in C and C++ when `break` statements are accidentally omitted. For example, the sections of the `switch` statement above can be reversed without affecting the behavior of the statement:
+The *statement_list*s of a *switch_block* may contain declaration statements ([§12.6](statements.md#126-declaration-statements)). The scope of a local variable or constant declared in a switch block is the switch block.
+
+Within a switch block, the meaning of a name used in an expression context shall always be the same ([§11.7.4](expressions.md#1174-simple-names)).
+
+The statement list of a given switch section is reachable if the `switch` statement is reachable and at least one of the following is true:
+
+- The switch expression value matches a `case` label in the switch section.
+- The switch expression value doesn't match any `case` label, and the switch section contains the `default` label.
+- A switch label of the switch section is referenced by a reachable `goto case` or `goto default` statement.
+
+The end point of a `switch` statement is reachable if at least one of the following is true:
+
+- The `switch` statement contains a reachable `break` statement that exits the `switch` statement.
+- The `switch` statement is reachable, the switch expression is a non-constant value, and no `default` label is present.
+- The `switch` statement is reachable, the switch expression is a constant value that doesn’t match any `case` label, and no `default` label is present.
+
+If a `case` label is unreachable, a compile-time occurs. (However, it is *not* a compile-time error for a `default` label to be unreachable).
```

Comment:
We need rules to be given for a `case` label to be reachable or unreachable.  We will need something saying that it is an error if a case pattern is subsumed by previous patterns.  See [HERE](https://sharplab.io/#v2:CYLg1APgAgTABAQQLACgDeq5blALHAWQAoAjAezIBs4BDAGjnKsYEpNsMVtu4BnAdwCWAFwDGACzhF6rOJx4K4omrwCmU4QCcArqoYB9FiHaLFJTapoBrANwnTSleqL6GW3UfsPs5y7a+KympSAGY0lGoMYRGqnlzePL7WdvHeQc6ucIYgcAD0uXCqmppkmjkAKuLqAiISjsGCvHDaAHYWNBI0JJSqAd5J/qk8AL72oyjjQA).  The behavior of this must be specified.  I'm afraid the margin here is too small to contain the specification.

Reply:
It is also worth saying "The order in which patterns are matched is not defined. A compiler is permitted to match patterns out of order, and to reuse the results of already matched patterns to compute the result of matching of other patterns."

Although this fact may not be observable in C# 7 except by inspecting the generated code or measuring performance, it will have real semantic consequences in future versions.  This is worth adding today so it doesn't get lost.

Reply:
Hmm, matching in any order with non-idempotent patterns... here be dragons! Can we change Rosyln so the order of evaluation is randomised on compile? :-)

Reply:
When we get to nested patterns, it will be recommended that properties tested in patterns be idempotent.